File: 1402-count-square-submatrices-with-all-ones/count-square-submatrices-with-all-ones.py
class Solution:
    def countSquares(self, matrix: List[List[int]]) -> int:
        ROW = len(matrix)
        COL = len(matrix[0])
        memo = {}

        def dfs(i, j):
            if (i, j) in memo: return memo[(i, j)]
            if i == ROW or j == COL or not matrix[i][j]:
                return 0
            memo[(i, j)] = 1 + min(dfs(i + 1, j), dfs(i, j + 1), dfs(i + 1, j + 1))
            return memo[(i, j)]
            
        result = 0
        for i in range(ROW):
            for j in range(COL):
                result += dfs(i, j)

        return result
        # ROW = len(matrix)
        # COL = len(matrix[0])
        # memo = defaultdict(int)
        # for i in range(ROW):
        #     for j in range(COL):
        #         if matrix[i][j] != 0:
        #             a, b, c = 0, 0, 0
        #             if i > 0:
        #                 a = matrix[i - 1][j]
        #             if j > 0:
        #                 b = matrix[i][j - 1]
        #             if i > 0 and j > 0:
        #                 c = matrix[i - 1][j - 1]
        #             matrix[i][j] = matrix[i][j] + min(a, b, c)
        #             for k in range(1, matrix[i][j] + 1):
        #                 memo[k] += 1
        # return sum(list(memo.values()))
        

File: 1402-count-square-submatrices-with-all-ones/count-square-submatrices-with-all-ones.java
class Solution {
    //bottom - 2 ROW
    public int countSquares(int[][] matrix) {
        int ROW = matrix.length, COL = matrix[0].length, result = 0;
        HashMap<Integer, Integer> memo1 = new HashMap<Integer, Integer>();
        for (int c = 0; c < COL; c++){
            memo1.put(c, 0);
        }
        for (int r = 0; r < ROW; r++){
            HashMap<Integer, Integer> memo2 = new HashMap<Integer, Integer>();
            for (int c = 0; c < COL; c++){
                if (c > 0 && matrix[r][c] == 1){
                    memo2.put(c, 1 + Math.min(memo1.get(c), Math.min(memo2.get(c - 1), memo1.get(c - 1))));
                    result += memo2.get(c);
                }
                else{
                    memo2.put(c, matrix[r][c]);
                    result += memo2.get(c);
                }
            }
            memo1 = memo2;
        }
        return result;
    }
    // bottom up
    // public int countSquares(int[][] matrix) {
    //     int ROW = matrix.length, COL = matrix[0].length, result = 0;
    //     HashMap<String, Integer> memo = new HashMap<String, Integer>();
    //     for (int r = 0; r < ROW; r++)
    //         for (int c = 0; c < COL; c++){
    //             if (r > 0 && c > 0 && matrix[r][c] != 0){
    //                 memo.put(String.valueOf(r) + "_" + String.valueOf(c), 1 + Math.min(memo.get(String.valueOf(r - 1) + "_" + String.valueOf(c)),
    //                 Math.min(memo.get(String.valueOf(r) + "_" + String.valueOf(c - 1)), memo.get(String.valueOf(r - 1) + "_" + String.valueOf(c - 1)))));
    //                 result +=  memo.get(String.valueOf(r) + "_" + String.valueOf(c));
    //             }
    //             else{
    //                 memo.put(String.valueOf(r) + "_" + String.valueOf(c), matrix[r][c]);
    //                 result += matrix[r][c];
    //             }             
    //         }
    //     return result;
    // }
    // dfs
    // int ROW = 0, COL = 0;

    // int dfs(int r, int c, int[][] matrix, HashMap<String, Integer> memo){
    //     String key = String.valueOf(r) + "_" + String.valueOf(c);
    //     if (memo.containsKey(key)) return memo.get(key);
    //     if (r == ROW || c == COL || matrix[r][c] == 0) return 0;
    //     memo.put(key, 1 + Math.min(dfs(r + 1, c, matrix, memo), Math.min(dfs(r, c + 1, matrix, memo), dfs(r + 1, c + 1, matrix, memo))));
    //     return memo.get(key);
    // }

    // public int countSquares(int[][] matrix) {
    //     int result = 0;
    //     ROW = matrix.length;
    //     COL = matrix[0].length;
    //     HashMap<String, Integer> memo = new HashMap<String, Integer>();
    //     for (int r = 0; r < ROW; r++)
    //         for (int c = 0; c < COL; c++)
    //             result += dfs(r, c, matrix, memo);
    //     return result;
    // }
}